src/opera_disp_tms/generate_sw_vel_tile.py

```diff
@@ -80,47 +64,49 @@ def parallel_linear_regression(x: np.ndarray, y: np.ndarray) -> np.ndarray:
     return slope_array
 
 
-def align_to_common_reference_date(granule_xrs: list[xr.DataArray], start_date: datetime) -> None:
-    granule_xrs.sort(key=lambda x: x.secondary_date)
-
-    if start_date <= granule_xrs[0].reference_date:
-        zero_xr = granule_xrs[0].copy()
-        zero_xr.attrs['secondary_date'] = zero_xr.attrs['reference_date']
-        zero_xr.data = np.zeros(zero_xr.shape)
-        granule_xrs.insert(0, zero_xr)
+def get_years_since_start(datetimes: List[datetime]) -> np.ndarray:
+    """Get the number of years since the earliest date as an array of floats.
+    Order of the datetimes is preserved.
 
-    previous_granule_xr = granule_xrs[0]
-    correction = -previous_granule_xr.data
-    for granule_xr in granule_xrs:
-        if not within_one_day(granule_xr.reference_date, previous_granule_xr.reference_date):
-            correction = previous_granule_xr.data
-        granule_xr += correction
-        previous_granule_xr = granule_xr
+    Args:
+        datetimes: A list of datetime objects
 
-    for granule_xr in granule_xrs:
-        granule_xr.attrs['reference_date'] = granule_xrs[0].secondary_date
+    Returns:
+        np.ndarray: The number of years since the earliest date as a list of floats
+    """
+    start = min(datetimes)
+    yrs_since_start = np.array([(date - start).days / 365.25 for date in datetimes])
+    return yrs_since_start
 
 
 def add_velocity_data_to_array(
-    granules: Iterable[Granule],
+    frame_id: int,
     geotransform: Affine,
+    begin_date: datetime,
+    end_date: datetime,
+    secant: bool,
     frame_map_array: np.ndarray,
     out_array: np.ndarray,
 ) -> np.ndarray:
-    """Create and add velocity data to an array using granules source from a single frame.
+    """Create and add velocity data to an array using granules sourced from a single frame.
 
     Args:
-        granules: A list of granule objects
+        frame_id: The frame id to use
         geotransform: The geotransform of the frame
+        begin_date: The start of the date range
+        end_date: The end of the date range
+        secant: Use a secant fit instead of a full linear fit
         frame_map_array: The frame map as a numpy array
         out_array: The array to add the velocity data to
 
     Returns:
         np.ndarray: The updated array
     """
     bbox = create_buffered_bbox(geotransform.to_gdal(), frame_map_array.shape, 90)  # EPSG:3857 is in meters
-    granule_xrs = [sw_disp.load_sw_disp_granule(x, bbox) for x in granules]
-    align_to_common_reference_date(granule_xrs, min(g.reference_date for g in granule_xrs))
+    strategy = 'spanning' if secant else 'all'
+    granule_xrs = load_sw_disp_stack(frame_id, bbox, begin_date, end_date, strategy)
+    if strategy == 'secant':
+        granule_xrs = [granule_xrs[0], granule_xrs[-1]]
```

Comment:
Again, we could be more memory-efficient here in the secant velocity case, but no need to optimize right now. The linear regression velocity case will have an order of magnitude higher memory demand, so any environment we run linear regression velocity in will have more than enough to handle the inefficient secant velocity case.